deposit: MintedDOIValidator fix

bouzlibop opened the pull request:
* FIX Fixes MintedDOIValidator, so that it correctly checks if DOI was
  already minted for the specific upload. (addresses zenodo/zenodo#285)

Signed-off-by: Adrian Pawel Baran <adrian.pawel.baran@cern.ch>

bouzlibop commented:
@lnielsen this is kind of radical change to `MintedDOIValidator`, please take a look.
I thought about getting rid of checking whether it's 'our' doi or not, but instead only consider differences between `field.data` and `field.object_data`.

lnielsen commented on invenio/modules/deposit/testsuite/test_deposit_validators.py:
```diff
@@ -0,0 +1,92 @@
+    pass
+
+
+ZENODO_DOI_PREFIX = "10.5072"
```
Can you use e.g. ``LOCAL_DOI_PREFIX`` and ``REMOTE_DOI_PREFIX`` instead. Better to not have mentions of Zenodo in Invenio code.

lnielsen commented:
Thanks a lot for adding the test. It looks good, but I just want to take an extra look just to make sure.

bouzlibop commented:
Sure, I've just amended the names (according to your comment above) and fixed `kwalitee` errors.
If it will be the case, that we don't need to have separate validations for local and remote dois, I guess, I will also remove references to prefix in init code of `MintedDOIValidator`.

jirikuncar commented on invenio/modules/deposit/validation_utils.py:
```diff
@@ -226,11 +240,12 @@ def __call__(self, form, field):
-        """
+        """Initialize validator.
+
         @param doi_prefix: DOI prefix, e.g. 10.5072
```
`s/@param /:param/`

jirikuncar commented:
@lnielsen are you ok with the change?

jirikuncar commented on invenio/modules/deposit/testsuite/test_deposit_validators.py:
```diff
@@ -0,0 +1,92 @@
+        from invenio.modules.deposit.validation_utils import MintedDOIValidator
+
+        validator = MintedDOIValidator()
+        field = Field("", LOCAL_DOI_PREFIX + "/zenodo.77777")
```
`s/zenodo/example` or `s/zenodo/test/`

jirikuncar commented:
@lnielsen shall we merge it to 2.0.5?

bouzlibop commented:
I've just updated this PR. Basically I added [this line](https://github.com/inveniosoftware/invenio/pull/3202/files#diff-ba026b84a5c248720661de03c70296deR269), as so I'm returning `StopValidation` instead of `None` in this validator.
I quickly tested it locally today, but then I guess we need to come up with some more extensive test cases for that.

bouzlibop commented:
@lnielsen, I've just rebased this one on top of the latest `master` branch. I think it is ready to go (as it was already integrated in `zenodo/invenio`) and as far as I know is working there.

I also noticed that there must to be done small change in `zenodo/invenio` (which is already added to this PR) in order to test be passing there:
```
24c24
< from wtforms.validators import ValidationError
---
> from wtforms.validators import StopValidation, ValidationError
53,54c53,56
<         self.assertIsNone(validator(form, field))
<         self.assertIsNone(validator(form, field2))
---
>         with self.assertRaises(StopValidation):
>             validator(form, field)
>         with self.assertRaises(StopValidation):
>             validator(form, field2)
68,69c70,73
<         self.assertIsNone(validator(form, field))
<         self.assertIsNone(validator(form, field2))
---
>         with self.assertRaises(StopValidation):
>             validator(form, field)
>         with self.assertRaises(StopValidation):
>             validator(form, field2)

```

bouzlibop commented:
moved to https://github.com/inveniosoftware/invenio-deposit/pull/2
